numcosmo/model/nc_hicosmo_sfb.c

```diff
@@ -0,0 +1,440 @@
+}
+
+/****************************************************************************
+ * x function, the scale factor today divided by the scale factor at a time \tau
+ ****************************************************************************/
+static gdouble
+_nc_hicosmo_sfb_x (NcHICosmo *cosmo, gdouble tau)
+{
+  gdouble x;
+  gdouble tabs = fabs (tau);
+  gdouble lnX_B = log (X_B);
+  x = exp(lnX_B - tabs);
+  
+  return x;
+}
+
+/****************************************************************************
+ * Normalized Hubble function redshift derivative
+ ****************************************************************************/
+static gdouble
+_nc_hicosmo_sfb_dE2_dt (NcHICosmo *cosmo, gdouble tau)
+{
+  const gdouble sign = GSL_SIGN (tau);
+  const gdouble lnX_B = log (X_B);
+  const gdouble tabs = fabs (tau);
+  const gdouble w = W;
+  const gdouble x_3_1pw = exp (3.0 * (1.0 + w) * (-tabs + lnX_B));
+  const gdouble Omega_w = OMEGA_W;
+  const gdouble e_t3 = exp (-3.0 *(1.0 - w) * tabs);
+
+  
+  return 3.0  * Omega_w * x_3_1pw * sign * e_t3 - 3 / 2 * Omega_w * (1.0 + w) * x_3_1pw * sign;
+}
+
+/****************************************************************************
+ * Speed of sound squared
+ ****************************************************************************/
+static gdouble
+_nc_hicosmo_sfb_bgp_cs2 (NcHICosmo *cosmo, gdouble tau)
+{
+  gdouble w = W;
+  
+  
+  return w;
+}
+
+
+static gdouble
+_nc_hicosmo_sfb_eval_z2 (NcHICosmo *cosmo, gdouble tau)
+{
+  gdouble x    = _nc_hicosmo_sfb_x (cosmo, tau);
+  gdouble w    = W;
+  gdouble mult = 3.0 * (1.0 + w) / (2.0 * w);
+  gdouble x3   = x * x * x;
+  gdouble E = sqrt (_nc_hicosmo_sfb_E2 (cosmo, tau));
+  gdouble N = 1.0 / E;
+  
+  return mult * 1.0 / (x3 * N);
+}
+
+/****************************************************************************
+ * Simple functions
+ ****************************************************************************/
+static gdouble
+_nc_hicosmo_sfb_H0 (NcHICosmo *cosmo)
+{
+  return MACRO_H0 * (OMEGA_R + OMEGA_W);
+}
+
+static gdouble
+_nc_hicosmo_sfb_Omega_t0 (NcHICosmo *cosmo)
+{
+  return 1.0;
+}
+
+static gdouble
+_nc_hicosmo_sfb_Omega_c0 (NcHICosmo *cosmo)
+{
+  return OMEGA_W;
+}
+
+static gdouble
+_nc_hicosmo_sfb_xb (NcHICosmo *cosmo)
+{
+  return X_B;
+}
+
+/*****************************************************************************
+ *  Interface functions
+ ******************************************************************************/
+static gdouble
+_nc_hipert_iadiab_eval_m (NcHIPertIAdiab *iad, const gdouble tau, const gdouble k)
+{
+  NcHICosmo *cosmo = NC_HICOSMO (iad);
+  gdouble z2    = _nc_hicosmo_sfb_eval_z2 (cosmo, tau);
+
+  return z2 * 2.0;
+} 
+ 
+ 
+static gdouble
+_nc_hipert_iadiab_eval_nu (NcHIPertIAdiab *iad, const gdouble tau, const gdouble k)
+{
+  NcHICosmo *cosmo = NC_HICOSMO (iad);
+  const gdouble w        = W;
+  const gdouble N        = 1.0 / sqrt (_nc_hicosmo_sfb_E2 (cosmo, tau));
```

Comment:
Compute all quantities in the same function for effectiveness

Reply:
done for all